hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/DomainSocketFactory.java

```diff
@@ -0,0 +1,268 @@
+ */
+public final class DomainSocketFactory implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(
+      DomainSocketFactory.class);
+
+  /**
+   *  Domain socket path state.
+   */
+  public enum PathState {
+    NOT_CONFIGURED(false),
+    DISABLED(false),
+    VALID(true);
+
+    PathState(boolean usableForShortCircuit) {
+      this.usableForShortCircuit = usableForShortCircuit;
+    }
+
+    public boolean getUsableForShortCircuit() {
+      return usableForShortCircuit;
+    }
+    private final boolean usableForShortCircuit;
+  }
+
+  /**
+   *  Domain socket path.
+   */
+  public static class PathInfo {
+    private static final PathInfo NOT_CONFIGURED = new PathInfo("", PathState.NOT_CONFIGURED);
+    private static final PathInfo DISABLED = new PathInfo("", PathState.DISABLED);
+    private static final PathInfo VALID = new PathInfo("", PathState.VALID);
+
+    private final String path;
+    private final PathState state;
+
+    PathInfo(String path, PathState state) {
+      this.path = path;
+      this.state = state;
+    }
+
+    public String getPath() {
+      return path;
+    }
+
+    public PathState getPathState() {
+      return state;
+    }
+
+    @Override
+    public String toString() {
+      return "PathInfo{path=" + path + ", state=" + state + "}";
+    }
+  }
+
+  public static final String FEATURE = "short-circuit reads";
+  public static final String FEATURE_FLAG = "SC";
+  private static boolean nativeCodeLoaded = false;
+  private static String nativeLibraryLoadFailureReason;
+  private long pathExpireMills;
+  private final ConcurrentHashMap<String, PathInfo> pathMap;
+  private Timer timer;
+  private boolean isEnabled = false;
+  private String domainSocketPath;
+
+  static {
+    // Try to load native hadoop library and set fallback flag appropriately
+    if (SystemUtils.IS_OS_WINDOWS) {
+      nativeLibraryLoadFailureReason = "UNIX Domain sockets are not available on Windows.";
+    } else {
+      LOG.info("Trying to load the custom-built native-hadoop library...");
+      try {
+        System.loadLibrary("hadoop");
+        LOG.info("Loaded the native-hadoop library");
+        nativeCodeLoaded = true;
+      } catch (Throwable t) {
+        // Ignore failure to continue
+        LOG.info("Failed to load native-hadoop with error: " + t);
+        LOG.info("java.library.path=" + System.getProperty("java.library.path"));
+        nativeLibraryLoadFailureReason = "libhadoop cannot be loaded.";
+      }
+
+      if (!nativeCodeLoaded) {
+        LOG.warn("Unable to load native-hadoop library for your platform... " +
+            "using builtin-java classes where applicable");
+      }
+    }
+  }
+
+  private static volatile DomainSocketFactory instance = null;
+
+  public static synchronized DomainSocketFactory getInstance(ConfigurationSource conf) {
+    if (instance == null) {
+      instance = new DomainSocketFactory(conf);
+    }
+    return instance;
+  }
```

Comment:
```suggestion
  public static DomainSocketFactory getInstance(ConfigurationSource conf) {
    if (instance == null) {
      synchronized (DomainSocketFactory.class) {
          instance = new DomainSocketFactory(conf);
      }
    }
    return instance;
  }
```

Use double-checked locking instead: https://www.baeldung.com/java-singleton-double-checked-locking